### ISSUE TITLE
CS4014 warning occurs when custom AsyncMethodBuilder without GetAwaiter

### ISSUE THREAD

**Issue by neuecc** [dotnet/roslyn#45476]
**Version Used**: 
Microsoft Visual Studio Enterprise 2019
Version 16.4.2
VisualStudio.16.Release/16.4.2+29613.14

![image](https://user-images.githubusercontent.com/46207/85833948-807bd800-b7cd-11ea-8fe7-72f62add1f3a.png)

`UniTaskVoid` is custom async type like async void, that does not implements GetAwaiter to prevent use with await.
It can not use await but CS4014 error suggests to use with await.
Currently we have to use `_ = FireAndForgetAsync();` to avoid warning.

Expected result is should not show warning when custom async type does not implement GetAwaiter.

This type is already released in OSS and used widely on Unity Game Engine.
https://github.com/Cysharp/UniTask/

Sample reproduce code

```csharp
using Cysharp.Threading.Tasks;
using Cysharp.Threading.Tasks.CompilerServices;
using System;
using System.Runtime.CompilerServices;
using System.Security;
using System.Threading.Tasks;

class Program
{
    static void Main(string[] args)
    {
        // show warning on this line.
        FireAndForgetAsync();
    }

    static async UniTaskVoid FireAndForgetAsync()
    {
        await Task.Yield();
    }
}

namespace Cysharp.Threading.Tasks
{
    [AsyncMethodBuilder(typeof(AsyncUniTaskVoidMethodBuilder))]
    public readonly struct UniTaskVoid
    {
    }
}


namespace Cysharp.Threading.Tasks.CompilerServices
{
    // Empty builder to easily reproduce(actural code has implementation):
    // https://github.com/Cysharp/UniTask/blob/master/src/UniTask/Assets/Plugins/UniTask/Runtime/CompilerServices/AsyncUniTaskVoidMethodBuilder.cs
    public struct AsyncUniTaskVoidMethodBuilder
    {
        public static AsyncUniTaskVoidMethodBuilder Create()
        {
            return default;
        }

        public UniTaskVoid Task
        {
            get
            {
                return default;
            }
        }

        public void SetException(Exception exception)
        {
        }

        public void SetResult()
        {
        }

        public void AwaitOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine)
            where TAwaiter : INotifyCompletion
            where TStateMachine : IAsyncStateMachine
        {
        }

        [SecuritySafeCritical]

        public void AwaitUnsafeOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine)
            where TAwaiter : ICriticalNotifyCompletion
            where TStateMachine : IAsyncStateMachine
        {
        }

        public void Start<TStateMachine>(ref TStateMachine stateMachine)
            where TStateMachine : IAsyncStateMachine
        {
            stateMachine.MoveNext();
        }

        public void SetStateMachine(IAsyncStateMachine stateMachine)
        {
        }
    }
}
```

**Comment by jaredpar** [comment]
Thanks for filing the issue but this behavior is "By Design" and I do not think that we should special case it in the manner you describe. This is not the standard intended usage of the feature. 

If the `UniTaskVoid` wants to avoid the warning the library can include a `DiagnosticSuppressor` to disable that warning.

**Comment by neuecc** [comment]
Thank you.
I don't disagree with Close, although I suggest that it's necessary for users to write async void (or something close to them).

Note that DiagnosticSuppressor does not work in Unity, at least not currently.

**Comment by jaredpar** [comment]
> Note that DiagnosticSuppressor does not work in Unity, at least not currently.

That is surprising to hear because they were one of the primary reasons that we implemented it. They had some good scenarios around serialization and field initialization warnings that eventually pushed us over the edge for doing the work.
